url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/MessagePack.java

```diff
@@ -0,0 +1,837 @@
+    }
+
+    /**
+     * Deserializes byte array to object of specified class.
+     * 
+     * @since 0.6.0
+     * @param bytes
+     *            input byte array
+     * @param c
+     * @return
+     * @throws IOException
+     */
+    public <T> T read(byte[] bytes, Class<T> c) throws IOException {
+        @SuppressWarnings("unchecked")
+        Template<T> tmpl = registry.lookup(c);
+        return read(bytes, null, tmpl);
+    }
+
+    /**
+     * Deserializes byte array to object according to specified template.
+     * 
+     * @since 0.6.0
+     * @param bytes
+     *            input byte array
+     * @param v
+     * @param tmpl
+     *            template
+     * @return
+     * @throws IOException
+     */
+    public <T> T read(byte[] bytes, T v, Template<T> tmpl) throws IOException {
+        BufferUnpacker u = createBufferUnpacker(bytes);
+        return (T) tmpl.read(u, v);
+    }
+
+    /**
+     * Deserializes byte array to object.
+     *
+     * @since 0.6.8
+     * @param bytes
+     *            input byte array
+     * @param v
+     * @return
+     * @throws IOException
+     */
+    public <T> T read(byte[] bytes, int off, int len, Class<T> c) throws IOException {
+        @SuppressWarnings("unchecked")
+        Template<T> tmpl = registry.lookup(c);
+        BufferUnpacker u = createBufferUnpacker(bytes, off, len);
+        return (T) tmpl.read(u, null);
+    }
+
+    /**
+     * Deserializes buffer to object.
+     * 
+     * @since 0.6.0
+     * @param b
+     *            input {@link java.nio.ByteBuffer} object
+     * @param v
+     * @return
+     * @throws IOException
+     */
+    public <T> T read(ByteBuffer b, T v) throws IOException {
+        @SuppressWarnings("unchecked")
+        Template<T> tmpl = registry.lookup(v.getClass());
+        return read(b, v, tmpl);
+    }
+
+    /**
+     * Deserializes buffer to object according to template.
+     * 
+     * @since 0.6.0
+     * @param b
+     *            input buffer object
+     * @param tmpl
+     * @return
+     * @throws IOException
+     */
+    public <T> T read(ByteBuffer b, Template<T> tmpl) throws IOException {
+        return read(b, null, tmpl);
+    }
+
+    /**
+     * Deserializes buffer to object of specified class.
+     * 
+     * @since 0.6.0
+     * @param b
+     * @param c
+     * @return
+     * @throws IOException
+     */
+    public <T> T read(ByteBuffer b, Class<T> c) throws IOException {
+        @SuppressWarnings("unchecked")
+        Template<T> tmpl = registry.lookup(c);
+        return read(b, null, tmpl);
+    }
+
+    /**
+     * Deserializes buffer to object according to template.
+     * 
+     * @since 0.6.0
+     * @param b
+     *            input buffer object
+     * @param v
+     * @param tmpl
+     * @return
+     * @throws IOException
+     */
+    public <T> T read(ByteBuffer b, T v, Template<T> tmpl) throws IOException {
+        BufferUnpacker u = createBufferUnpacker(b);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *u*. Currently, there are execution paths that do not contain closure statements, for example, when Template.read throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).